internal/xds/util.go

```diff
@@ -38,3 +63,161 @@ func ClusterLoadAssignmentName(service types.NamespacedName, portName string) st
+}
+
+func rewriteAnyMessage(a *any.Any) {
+	if a != nil {
+		anyval := ptypes.DynamicAny{}
+		if err := ptypes.UnmarshalAny(a, &anyval); err != nil {
+			panic(fmt.Sprintf("failed to unmarshal %T: %s", a, err.Error()))
+		}
+
+		newMsg := protobuf.MustMarshalAny(Rewrite(anyval.Message))
+		if replacement, ok := TypeMapping[newMsg.TypeUrl]; ok {
+			newMsg.TypeUrl = replacement
+		}
+
+		a.TypeUrl = newMsg.TypeUrl
+		a.Value = newMsg.Value
+	}
+}
+
+func rewriteConfigSource(s *envoy_api_core_v2.ConfigSource) {
+	if s != nil {
+		s.ResourceApiVersion = envoy_api_core_v2.ApiVersion_V3
+		s.GetApiConfigSource().TransportApiVersion = envoy_api_core_v2.ApiVersion_V3
+	}
+}
+
+// Rewrite changes the given xDS message to use the v3 xDS API.
+//
+// Since the v2 and v3 APIs are wire-compatible, we just rewrite
+// the type names for type URLs in any.Any messages. This allows Envoy
+// to do the actual conversion, and Envoy takes care of migrating
+// deprecated fields.
+func Rewrite(in proto.Message) proto.Message {
+	switch msg := in.(type) {
+	case *envoy_api_v2.ClusterLoadAssignment:
+		return msg
+	case *envoy_api_auth_v2.Secret:
+		return msg
+
+	case *envoy_api_v2.Cluster:
+		if e := msg.GetEdsClusterConfig(); e != nil {
+			rewriteConfigSource(e.GetEdsConfig())
+		}
+
+		if t := msg.GetTransportSocket(); t != nil {
+			rewriteAnyMessage(t.GetTypedConfig())
+		}
+
+		return msg
+
+	case *envoy_api_v2.RouteConfiguration:
+		for _, v := range msg.GetVirtualHosts() {
+			for _, r := range v.GetRoutes() {
+				for _, conf := range r.GetTypedPerFilterConfig() {
+					rewriteAnyMessage(conf)
+				}
+			}
+		}
+
+		return msg
+
+	case *envoy_api_v2.Listener:
+		for _, filter := range msg.ListenerFilters {
+			rewriteAnyMessage(filter.GetTypedConfig())
+		}
+
+		for _, chain := range msg.FilterChains {
+			for _, filter := range chain.Filters {
+				rewriteAnyMessage(filter.GetTypedConfig())
+			}
+
+			if t := chain.GetTransportSocket(); t != nil {
+				rewriteAnyMessage(t.GetTypedConfig())
+			}
+		}
+
+		for _, a := range msg.AccessLog {
+			rewriteAnyMessage(a.GetTypedConfig())
+		}
+
+		return msg
+
+	case *envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager:
+		if r := msg.GetRds(); r != nil {
+			rewriteConfigSource(r.GetConfigSource())
+		}
+
+		for _, f := range msg.HttpFilters {
+			rewriteAnyMessage(f.GetTypedConfig())
+		}
+
+		for _, l := range msg.AccessLog {
+			rewriteAnyMessage(l.GetTypedConfig())
+		}
+
+		return msg
+
+	case *envoy_api_auth_v2.DownstreamTlsContext:
+		for _, s := range msg.GetCommonTlsContext().TlsCertificateSdsSecretConfigs {
+			rewriteConfigSource(s.GetSdsConfig())
+		}
+
+		return msg
+
+	case *envoy_api_auth_v2.UpstreamTlsContext:
+		for _, s := range msg.GetCommonTlsContext().TlsCertificateSdsSecretConfigs {
+			rewriteConfigSource(s.GetSdsConfig())
+		}
+
+		return msg
+
+	default:
+		log.Printf("missing conversion for %T, good luck", msg)
```

Comment:
Add the context where this is logging and remove `good luck`, it's not needed.

Reply:
If we're gonna log anything here (not sure we need to once this is finalized), we should plumb a logrus logger through